## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


wholesales_df = pd.read_csv('Resources/client_dataset.csv')
wholesales_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
wholesales_df.shape 

(54639, 18)

In [3]:
# View the column names in the data

wholesales_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics

wholesales_df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# Check the shape of the dataframe (row and column count)
wholesales_df.dtypes  


# Get information about the dataframe, including data types and non-null counts
wholesales_df.info()  






<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [6]:
# View unique values in the 'category' column to identify different product categories
wholesales_df['category'].unique()  


array(['decor', 'consumables', 'software', 'furniture', 'electronics'],
      dtype=object)

In [7]:
# View unique values in the 'subcategory' column to identify different subcategories within the categories
wholesales_df['subcategory'].unique()  

array(['wall art', 'pens', 'project management', 'misc', 'plants',
       'kitchen supplies', 'paper products', 'seating', 'computers',
       'stands', 'tables', 'bathroom supplies', 'audio visual',
       'accounting', 'logistics', 'telecom', 'desks', 'cabling'],
      dtype=object)

In [8]:

# Check for any missing values in each column of the dataframe
wholesales_df.isnull().sum()  


first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64

In [9]:
# Count the occurrences of each category in the 'category' column to see which ones are the most frequent
wholesales_df['category'].value_counts()  

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

In [10]:
# View the first 5 rows of the dataframe to get a snapshot of the data
wholesales_df.head()  



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [11]:
# View the last 5 rows of the dataframe to inspect the end of the dataset
wholesales_df.tail()  

first       last                      job               phone  \
54634   Lauren      Reese           Radio producer  977.877.5272x11382   
54635  Derrick      Moore                 Musician        358.661.5483   
54636   Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637    Wanda    Solomon             Toxicologist       (311)767-4924   
54638   Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                    email  client_id  order_id  order_date  \
54634  late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635        derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636           gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637           solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                    belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  
54634            8  
54635            0  
54636            6  
54637            0  
54638            5

In [12]:
# What three item categories had the most entries?

# Find the three item categories with the most entries
top_three_categories = wholesales_df['category'].value_counts().head(3)

# Display the result
top_three_categories


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [13]:
# Step 1: Find the category with the most entries
most_common_category = wholesales_df['category'].value_counts().idxmax()

# Step 2: Filter the dataframe to only include rows where the category matches the most common category
filtered_df = wholesales_df[wholesales_df['category'] == most_common_category]

# Step 3: Find the subcategory with the most entries within the most common category
top_subcategory_in_most_common_category = filtered_df['subcategory'].value_counts().idxmax()

# Display the results using a formatted f-string
print(f"The category with the most entries is '{most_common_category}'")
print(f"The subcategory with the most entries within this category is '{top_subcategory_in_most_common_category}'")


The category with the most entries is 'consumables'
The subcategory with the most entries within this category is 'bathroom supplies'


In [14]:
# Which five clients had the most entries in the data?

# Step 1: Find the five clients with the most entries using client_id
top_five_clients = wholesales_df['client_id'].value_counts().head(5)

# Step 2: Display the result with client_id and their corresponding first and last names
for client_id in top_five_clients.index:
    client_data = wholesales_df[wholesales_df['client_id'] == client_id].iloc[0]
    print(f"Client ID: {client_id}, Name: {client_data['first']} {client_data['last']}, Entries: {top_five_clients[client_id]}")


Client ID: 33615, Name: Jessica Reyes, Entries: 220
Client ID: 66037, Name: Angela Everett, Entries: 211
Client ID: 46820, Name: Bryan Myers, Entries: 209
Client ID: 38378, Name: Alexandra Young, Entries: 207
Client ID: 24741, Name: Kendra Garrett, Entries: 207


In [15]:
# Store the client ids of those top 5 clients in a list.
top_five_client_ids = top_five_clients.index.tolist()

# Display the result
print("Top 5 client IDs:", top_five_client_ids)

Top 5 client IDs: [33615, 66037, 46820, 38378, 24741]


In [16]:
# How many total units (the qty column) did the client with the most entries order order?

# Step 1: Find the five clients with the most entries using client_id
top_five_clients = wholesales_df['client_id'].value_counts().head(5)

# Step 2: Store the client ids of those top 5 clients in a list
top_five_client_ids = top_five_clients.index.tolist()

# Step 3: Display the first and last names along with the client IDs
print("Top 5 Clients with Most Entries:")
for client_id in top_five_client_ids:
    # Extract first and last name for the client
    client_data = wholesales_df[wholesales_df['client_id'] == client_id].iloc[0]
    first_name = client_data['first']
    last_name = client_data['last']
    print(f"Client ID: {client_id}, Name: {first_name} {last_name}, Entries: {top_five_clients[client_id]}")

# Step 4: Find the client with the most entries
client_with_most_entries = top_five_client_ids[0]

# Step 5: Calculate the total units (qty) ordered by the client with the most entries
total_units_most_entries = wholesales_df[wholesales_df['client_id'] == client_with_most_entries]['qty'].sum()

# Output the total units using string formatting for thousands separator
print(f"\nTotal units ordered by the client with the most entries (Client ID: {client_with_most_entries}): {total_units_most_entries:,}")




Top 5 Clients with Most Entries:
Client ID: 33615, Name: Jessica Reyes, Entries: 220
Client ID: 66037, Name: Angela Everett, Entries: 211
Client ID: 46820, Name: Bryan Myers, Entries: 209
Client ID: 38378, Name: Alexandra Young, Entries: 207
Client ID: 24741, Name: Kendra Garrett, Entries: 207

Total units ordered by the client with the most entries (Client ID: 33615): 64,313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [17]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty

# Make a copy of the original dataframe to preserve the original data
updated_wholesales_df = wholesales_df.copy()

# Create a new column 'subtotal' that calculates the subtotal for each line
updated_wholesales_df['subtotal'] = updated_wholesales_df['unit_price'] * updated_wholesales_df['qty']

# Display the first few rows to check the new column
updated_wholesales_df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [18]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.


# Create a new column 'shipping_price' based on the condition of unit_weight
updated_wholesales_df['shipping_price'] = np.where(
    updated_wholesales_df['unit_weight'] > 50,  # Condition: if unit_weight is greater than 50
    updated_wholesales_df['unit_weight'] * 7,    # Shipping price is $7 per pound if the weight is greater than 50
    updated_wholesales_df['unit_weight'] * 10    # Shipping price is $10 per pound if the weight is 50 or under
)

# Display the first few rows to check the new 'shipping_price' column
updated_wholesales_df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0            75.0  
1            14.9  
2            16.8  
3            12.3  
4           464.3

In [19]:
# Check if there are any items with a unit weight greater than 50 lbs
items_over_50_lbs = updated_wholesales_df['unit_weight'] > 50

# Display the result (True if there are any, False if not)
items_over_50_lbs.any()


False

In [20]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

# Define the sales tax rate (convert percentage to a decimal)
sales_tax_rate = 9.25 / 100

# Create a new column 'total_price' that includes subtotal, shipping, and applies sales tax to the sum of both
updated_wholesales_df['total_price'] = (updated_wholesales_df['subtotal'] + updated_wholesales_df['shipping_price']) * (1 + sales_tax_rate)

# Display the first few rows to check the new 'total_price' column
updated_wholesales_df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_price  \
0     762.71         7.50  105            1  115164.00            75.0   
1      15.09         1.49   21            0     523.95            14.9   
2       7.86         1.68   39            6     527.28            16.8   
3      24.85         1.23   29            3    1056.18            12.3   
4     108.17        46.43   20            1    3902.00           464.3   

     total_price  
0  125898.607500  
1     588.693625  
2     594.407400  
3    1167.314400  
4    4770.182750  

[5 rows x 21 columns]

In [21]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).


# Create a new column 'line_cost' that calculates the total cost for each line
updated_wholesales_df['line_cost'] = (updated_wholesales_df['unit_cost'] * updated_wholesales_df['qty']) + updated_wholesales_df['shipping_price']

# Display the first few rows to check the new 'line_cost' column
updated_wholesales_df.head()




first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price    total_price  \
0         7.50  105            1  115164.00            75.0  125898.607500   
1         1.49   21            0     523.95            14.9     588.693625   
2         1.68   39            6     527.28            16.8     594.407400   
3         1.23   29            3    1056.18            12.3    1167.314400   
4        46.43   20            1    3902.00           464.3    4770.182750   

   line_cost  
0   80159.55  
1     331.79  
2     323.34  
3     732.95  
4    2627.70  

[5 rows x 22 columns]

In [22]:
updated_wholesales_df.dtypes

first              object
last               object
job                object
phone              object
email              object
client_id           int64
order_id            int64
order_date         object
order_week          int64
order_year          int64
item_id            object
category           object
subcategory        object
unit_price        float64
unit_cost         float64
unit_weight       float64
qty                 int64
line_number         int64
subtotal          float64
shipping_price    float64
total_price       float64
line_cost         float64
dtype: object

In [23]:
updated_wholesales_df['order_date'] = pd.to_datetime(updated_wholesales_df['order_date'])


In [24]:
updated_wholesales_df.dtypes

first                     object
last                      object
job                       object
phone                     object
email                     object
client_id                  int64
order_id                   int64
order_date        datetime64[ns]
order_week                 int64
order_year                 int64
item_id                   object
category                  object
subcategory               object
unit_price               float64
unit_cost                float64
unit_weight              float64
qty                        int64
line_number                int64
subtotal                 float64
shipping_price           float64
total_price              float64
line_cost                float64
dtype: object

In [25]:
# Create a column for the profit of each line using line cost and line price

# Make a copy of the updated DataFrame for the profit calculation
profit_wholesales_df = updated_wholesales_df.copy()

# Create a new column 'profit' that calculates the profit for each line
profit_wholesales_df['profit'] = profit_wholesales_df['total_price'] - profit_wholesales_df['line_cost']

# Display the first few rows to check the new 'profit' column
profit_wholesales_df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482 2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089 2023-05-19   
2                swatson8146@payne.net      57113   1902144 2023-01-29   
3               bakernathan@benson.com      46554   9031802 2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274 2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price    total_price  line_cost  \
0            1  115164.00            75.0  125898.607500   80159.55   
1            0     523.95            14.9     588.693625     331.79   
2            6     527.28            16.8     594.407400     323.34   
3            3    1056.18            12.3    1167.314400     732.95   
4            1    3902.00           464.3    4770.182750    2627.70   

         profit  
0  45739.057500  
1    256.903625  
2    271.067400  
3    434.364400  
4   2142.482750  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [26]:
# Group the DataFrame by 'order_id' and sum the 'total_price' for each order
grouped_totals = profit_wholesales_df.groupby('order_id')['total_price'].sum().reset_index()

# Store the expected totals for the 3 orders from the receipts
expected_totals = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

# Loop through each order ID and compare the calculated total price with the expected total
for order_id, expected_total in expected_totals.items():
    # Get the calculated total from the grouped data
    calculated_total = grouped_totals[grouped_totals['order_id'] == order_id]['total_price'].values[0]
    
    # Print the result of the comparison
    print(f"Order ID: {order_id}")
    print(f"Calculated Total Price: {calculated_total:.2f}")
    print(f"Expected Total Price: {expected_total:.2f}")
    
    # Check if the calculated total matches the expected total
    if round(calculated_total, 2) == expected_total:
        print(f"Total price for Order ID {order_id} matches the expected total.\n")
    else:
        print(f"Total price for Order ID {order_id} does NOT match the expected total.\n")


Order ID: 2742071
Calculated Total Price: 139844.51
Expected Total Price: 152811.89
Total price for Order ID 2742071 does NOT match the expected total.

Order ID: 2173913
Calculated Total Price: 119838.36
Expected Total Price: 162388.71
Total price for Order ID 2173913 does NOT match the expected total.

Order ID: 6128929
Calculated Total Price: 752748.75
Expected Total Price: 923441.25
Total price for Order ID 6128929 does NOT match the expected total.



In [27]:
# Loop through each order ID and break down the components of total_price
for order_id, expected_total in expected_totals.items():
    # Filter the DataFrame for the given order ID
    order_df = profit_wholesales_df[profit_wholesales_df['order_id'] == order_id]
    
    # Sum the subtotal and shipping_price for the order
    subtotal_sum = order_df['subtotal'].sum()
    shipping_sum = order_df['shipping_price'].sum()
    
    # Recalculate the total price by applying sales tax on subtotal + shipping
    sales_tax_rate = 9.25 / 100
    recalculated_total_price = (subtotal_sum + shipping_sum) * (1 + sales_tax_rate)
    
    # Get the actual total price from the DataFrame
    actual_total_price = order_df['total_price'].sum()
    
    # Display the components and the recalculated total
    print(f"Order ID: {order_id}")
    print(f"Subtotal Sum: {subtotal_sum:.2f}")
    print(f"Shipping Price Sum: {shipping_sum:.2f}")
    print(f"Recalculated Total Price: {recalculated_total_price:.2f}")
    print(f"Actual Total Price in Data: {actual_total_price:.2f}")
    print(f"Expected Total Price: {expected_total:.2f}")
    
    # Check if the recalculated total matches the expected total
    if round(recalculated_total_price, 2) == expected_total:
        print(f"Recalculated total price for Order ID {order_id} matches the expected total.\n")
    else:
        print(f"Recalculated total price for Order ID {order_id} does NOT match the expected total.\n")


Order ID: 2742071
Subtotal Sum: 127799.33
Shipping Price Sum: 204.80
Recalculated Total Price: 139844.51
Actual Total Price in Data: 139844.51
Expected Total Price: 152811.89
Recalculated total price for Order ID 2742071 does NOT match the expected total.

Order ID: 2173913
Subtotal Sum: 109440.76
Shipping Price Sum: 251.10
Recalculated Total Price: 119838.36
Actual Total Price in Data: 119838.36
Expected Total Price: 162388.71
Recalculated total price for Order ID 2173913 does NOT match the expected total.

Order ID: 6128929
Subtotal Sum: 688564.27
Shipping Price Sum: 450.60
Recalculated Total Price: 752748.75
Actual Total Price in Data: 752748.75
Expected Total Price: 923441.25
Recalculated total price for Order ID 6128929 does NOT match the expected total.



In [28]:
# Loop through each order ID and verify the total weight and corresponding shipping price
for order_id, expected_total in expected_totals.items():
    # Filter the DataFrame for the given order ID
    order_df = profit_wholesales_df[profit_wholesales_df['order_id'] == order_id]
    
    # Calculate the total weight for the order
    total_weight = order_df['unit_weight'].sum()
    
    # Calculate the expected shipping price based on the total weight
    if total_weight > 50:
        expected_shipping_price = total_weight * 7  # $7 per pound if total weight > 50 lbs
    else:
        expected_shipping_price = total_weight * 10  # $10 per pound if total weight <= 50 lbs
    
    # Get the actual shipping price from the DataFrame
    actual_shipping_price = order_df['shipping_price'].sum()
    
    # Display the total weight and shipping price comparison
    print(f"Order ID: {order_id}")
    print(f"Total Weight: {total_weight:.2f} lbs")
    print(f"Expected Shipping Price: {expected_shipping_price:.2f}")
    print(f"Actual Shipping Price: {actual_shipping_price:.2f}")
    
    # Check if the shipping price matches the expected shipping price
    if round(expected_shipping_price, 2) == round(actual_shipping_price, 2):
        print(f"Shipping price for Order ID {order_id} matches the expected price.\n")
    else:
        print(f"Shipping price for Order ID {order_id} does NOT match the expected price.\n")



Order ID: 2742071
Total Weight: 20.48 lbs
Expected Shipping Price: 204.80
Actual Shipping Price: 204.80
Shipping price for Order ID 2742071 matches the expected price.

Order ID: 2173913
Total Weight: 25.11 lbs
Expected Shipping Price: 251.10
Actual Shipping Price: 251.10
Shipping price for Order ID 2173913 matches the expected price.

Order ID: 6128929
Total Weight: 45.06 lbs
Expected Shipping Price: 450.60
Actual Shipping Price: 450.60
Shipping price for Order ID 6128929 matches the expected price.



In [29]:
# Count the number of duplicate order_id entries
num_duplicates = profit_wholesales_df['order_id'].duplicated().sum()

# Print the result
print(f"Number of duplicate order IDs found: {num_duplicates}")



Number of duplicate order IDs found: 44639


In [30]:
 unique_order_count = profit_wholesales_df['order_id'].nunique()

print(f"the number of individual orders is: {unique_order_count:,}")

the number of individual orders is: 10,000


In [31]:

# Confirm that your calculations match the receipts. Remember, each order has multiple lines.

# Group the DataFrame by 'order_id' and sum the 'total_price' for each order
grouped_totals = profit_wholesales_df.groupby('order_id')['total_price'].sum().reset_index()

# Store the expected totals for the 3 orders from the receipts
expected_totals = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

# Loop through each order ID and compare the calculated total price with the expected total
for order_id, expected_total in expected_totals.items():
    # Get the calculated total from the grouped data
    calculated_total = grouped_totals[grouped_totals['order_id'] == order_id]['total_price'].values[0]
    
    # Print the result of the comparison
    print(f"Order ID: {order_id}")
    print(f"Calculated Total Price: {calculated_total:.2f}")
    print(f"Expected Total Price: {expected_total:.2f}")
    
    # Check if the calculated total matches the expected total
    if round(calculated_total, 2) == expected_total:
        print(f"Total price for Order ID {order_id} matches the expected total.\n")
    else:
        print(f"Total price for Order ID {order_id} does NOT match the expected total.\n")







Order ID: 2742071
Calculated Total Price: 139844.51
Expected Total Price: 152811.89
Total price for Order ID 2742071 does NOT match the expected total.

Order ID: 2173913
Calculated Total Price: 119838.36
Expected Total Price: 162388.71
Total price for Order ID 2173913 does NOT match the expected total.

Order ID: 6128929
Calculated Total Price: 752748.75
Expected Total Price: 923441.25
Total price for Order ID 6128929 does NOT match the expected total.



## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [32]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

# Step 1: Find the Top 5 Clients by Quantity
top_5_clients_by_qty = profit_wholesales_df.groupby('client_id')['qty'].sum().nlargest(5).reset_index()

# Step 2: Calculate the Total Spending for Each Client
top_5_clients_spending = profit_wholesales_df.groupby('client_id')['total_price'].sum().reset_index()

# Step 3: Filter to show only the top 5 clients by quantity
top_5_clients_spending = top_5_clients_spending[top_5_clients_spending['client_id'].isin(top_5_clients_by_qty['client_id'])]

# Step 4: Merge total quantity with the spending data
top_5_clients_spending = top_5_clients_spending.merge(top_5_clients_by_qty, on='client_id')

# Step 5: Merge with the original DataFrame to add client names
top_5_clients_spending = top_5_clients_spending.merge(profit_wholesales_df[['client_id', 'first', 'last']].drop_duplicates(), on='client_id')

# Step 6: Format the total_price as currency
top_5_clients_spending['total_price'] = top_5_clients_spending['total_price'].apply(lambda x: f"${x:,.2f}")

# Step 7: Display the results with client names and total quantity
print("Top 5 Clients by Quantity and Their Total Spending:")
print(top_5_clients_spending[['first', 'last', 'client_id', 'qty', 'total_price']])






Top 5 Clients by Quantity and Their Total Spending:
     first      last  client_id      qty      total_price
0  Randall    Watson      30143   846069   $23,518,475.21
1   Amanda   Burnett      77354   506382   $45,927,908.98
2    Julie  Anderson      78965  4033302  $650,096,959.42
3   Dennis   Hawkins      85931   644423  $226,361,135.01
4      Amy   Stewart      86721  1010426  $110,986,238.44


In [33]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

# Step 1: Find the Top 5 Clients by Total Units Purchased (Quantity)
top_5_clients_by_qty = profit_wholesales_df.groupby('client_id')['qty'].sum().nlargest(5).reset_index()

# Step 2: Aggregate the required metrics for the top 5 clients
summary_df = profit_wholesales_df.groupby('client_id').agg({
    'qty': 'sum',                    # Total units purchased
    'shipping_price': 'sum',          # Total shipping price
    'total_price': 'sum',             # Total revenue
    'profit': 'sum'                   # Total profit
}).reset_index()

# Step 3: Filter the summary_df to only include the top 5 clients by quantity
summary_df = summary_df[summary_df['client_id'].isin(top_5_clients_by_qty['client_id'])]

# Step 4: Merge with client names from the original DataFrame
summary_df = summary_df.merge(profit_wholesales_df[['client_id', 'first', 'last']].drop_duplicates(), on='client_id')

# Step 5: Format the total_price and profit as currency
summary_df['total_price'] = summary_df['total_price'].apply(lambda x: f"${x:,.2f}")
summary_df['profit'] = summary_df['profit'].apply(lambda x: f"${x:,.2f}")

# Step 6: Display the final summary DataFrame
print("Summary for the Top 5 Clients:")
summary_df[['first', 'last', 'client_id', 'qty', 'shipping_price', 'total_price', 'profit']]


Summary for the Top 5 Clients:


first      last  client_id      qty  shipping_price      total_price  \
0  Randall    Watson      30143   846069          2157.3   $23,518,475.21   
1   Amanda   Burnett      77354   506382          2564.3   $45,927,908.98   
2    Julie  Anderson      78965  4033302          7383.3  $650,096,959.42   
3   Dennis   Hawkins      85931   644423          2763.0  $226,361,135.01   
4      Amy   Stewart      86721  1010426          6154.6  $110,986,238.44   

            profit  
0   $10,419,759.21  
1   $10,229,638.87  
2  $246,859,505.41  
3  $108,358,048.60  
4   $37,024,209.98

In [34]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 

# Step 1: Define a function to convert dollar amounts to millions
def currency_format_millions(amount):
    return f"${amount / 1_000_000:.2f}M"

# Step 2: Identify the money columns (total_price and profit)
money_columns = ['total_price', 'profit']

# Step 3: Ensure the columns are numeric and apply the currency_format_millions function
for col in money_columns:
    summary_df[col] = pd.to_numeric(summary_df[col], errors='coerce')  # Convert to numeric if needed
    summary_df[col] = summary_df[col].apply(currency_format_millions)

# Step 4: Rename the columns for presentation
summary_df.rename(columns={
    'qty': 'Total Units Purchased',
    'shipping_price': 'Total Shipping Price',
    'total_price': 'Total Revenue (in millions)',
    'profit': 'Total Profit (in millions)'
}, inplace=True)

# Step 5: Display the formatted DataFrame with renamed columns
print("Formatted Summary for the Top 5 Clients:")
summary_df[['first', 'last', 'client_id', 'Total Units Purchased', 'Total Shipping Price', 'Total Revenue (in millions)', 'Total Profit (in millions)']]









Formatted Summary for the Top 5 Clients:


first      last  client_id  Total Units Purchased  Total Shipping Price  \
0  Randall    Watson      30143                 846069                2157.3   
1   Amanda   Burnett      77354                 506382                2564.3   
2    Julie  Anderson      78965                4033302                7383.3   
3   Dennis   Hawkins      85931                 644423                2763.0   
4      Amy   Stewart      86721                1010426                6154.6   

  Total Revenue (in millions) Total Profit (in millions)  
0                       $nanM                      $nanM  
1                       $nanM                      $nanM  
2                       $nanM                      $nanM  
3                       $nanM                      $nanM  
4                       $nanM                      $nanM

In [35]:
summary_df.dtypes

client_id                        int64
Total Units Purchased            int64
Total Shipping Price           float64
Total Revenue (in millions)     object
Total Profit (in millions)      object
first                           object
last                            object
dtype: object

In [36]:
summary_df.isnull().sum()

client_id                      0
Total Units Purchased          0
Total Shipping Price           0
Total Revenue (in millions)    0
Total Profit (in millions)     0
first                          0
last                           0
dtype: int64

In [37]:
# Step 1: Convert "Total Profit (in millions)" back to numeric for sorting
summary_df['Total Profit (in millions)'] = summary_df['Total Profit (in millions)'].replace({'\$': '', 'M': ''}, regex=True).astype(float)

# Step 2: Sort the DataFrame by "Total Profit (in millions)" from highest to lowest
sorted_df = summary_df.sort_values(by='Total Profit (in millions)', ascending=False)

# Step 3: Display the sorted DataFrame
print("Sorted Summary by Total Profit (in millions):")
sorted_df[['first', 'last', 'client_id', 'Total Units Purchased', 'Total Shipping Price', 'Total Revenue (in millions)', 'Total Profit (in millions)']]


Sorted Summary by Total Profit (in millions):


first      last  client_id  Total Units Purchased  Total Shipping Price  \
0  Randall    Watson      30143                 846069                2157.3   
1   Amanda   Burnett      77354                 506382                2564.3   
2    Julie  Anderson      78965                4033302                7383.3   
3   Dennis   Hawkins      85931                 644423                2763.0   
4      Amy   Stewart      86721                1010426                6154.6   

  Total Revenue (in millions)  Total Profit (in millions)  
0                       $nanM                         NaN  
1                       $nanM                         NaN  
2                       $nanM                         NaN  
3                       $nanM                         NaN  
4                       $nanM                         NaN

In [38]:
# Remove rows where 'Total Profit (in millions)' is NaN
summary_df_cleaned = summary_df.dropna(subset=['Total Profit (in millions)'])

# Sort the cleaned DataFrame by "Total Profit (in millions)" from highest to lowest
sorted_df = summary_df_cleaned.sort_values(by='Total Profit (in millions)', ascending=False)

# Display the sorted DataFrame
print("Sorted Summary by Total Profit (in millions):")
sorted_df[['first', 'last', 'client_id', 'Total Units Purchased', 'Total Shipping Price', 'Total Revenue (in millions)', 'Total Profit (in millions)']]


Sorted Summary by Total Profit (in millions):


Empty DataFrame
Columns: [first, last, client_id, Total Units Purchased, Total Shipping Price, Total Revenue (in millions), Total Profit (in millions)]
Index: []

In [39]:
# Fill NaN values in 'Total Profit (in millions)' with 0
summary_df['Total Profit (in millions)'] = summary_df['Total Profit (in millions)'].fillna(0)

# Sort the DataFrame by "Total Profit (in millions)" from highest to lowest
sorted_df = summary_df.sort_values(by='Total Profit (in millions)', ascending=False)

# Display the sorted DataFrame
print("Sorted Summary by Total Profit (in millions):")
sorted_df[['first', 'last', 'client_id', 'Total Units Purchased', 'Total Shipping Price', 'Total Revenue (in millions)', 'Total Profit (in millions)']]


Sorted Summary by Total Profit (in millions):


first      last  client_id  Total Units Purchased  Total Shipping Price  \
0  Randall    Watson      30143                 846069                2157.3   
1   Amanda   Burnett      77354                 506382                2564.3   
2    Julie  Anderson      78965                4033302                7383.3   
3   Dennis   Hawkins      85931                 644423                2763.0   
4      Amy   Stewart      86721                1010426                6154.6   

  Total Revenue (in millions)  Total Profit (in millions)  
0                       $nanM                         0.0  
1                       $nanM                         0.0  
2                       $nanM                         0.0  
3                       $nanM                         0.0  
4                       $nanM                         0.0

In [40]:
# Fill NaN values in 'Total Profit (in millions)' with the mean of the column
summary_df['Total Profit (in millions)'] = summary_df['Total Profit (in millions)'].fillna(summary_df['Total Profit (in millions)'].mean())

# Sort the DataFrame by "Total Profit (in millions)" from highest to lowest
sorted_df = summary_df.sort_values(by='Total Profit (in millions)', ascending=False)

# Display the sorted DataFrame
print("Sorted Summary by Total Profit (in millions):")
sorted_df[['first', 'last', 'client_id', 'Total Units Purchased', 'Total Shipping Price', 'Total Revenue (in millions)', 'Total Profit (in millions)']]


Sorted Summary by Total Profit (in millions):


first      last  client_id  Total Units Purchased  Total Shipping Price  \
0  Randall    Watson      30143                 846069                2157.3   
1   Amanda   Burnett      77354                 506382                2564.3   
2    Julie  Anderson      78965                4033302                7383.3   
3   Dennis   Hawkins      85931                 644423                2763.0   
4      Amy   Stewart      86721                1010426                6154.6   

  Total Revenue (in millions)  Total Profit (in millions)  
0                       $nanM                         0.0  
1                       $nanM                         0.0  
2                       $nanM                         0.0  
3                       $nanM                         0.0  
4                       $nanM                         0.0